docs(api): liquid presence verification

jwwojak opened the pull request:
# Overview

Adding new documentation to describe new Liquid Presence Verification methods, includes updating docstrings.  Sandbox links below.

The JIRA ticket is [RTC-148](https://opentrons.atlassian.net/browse/RTC-148).


# Test Plan

Review the new docs.

# Changelog

**New**
Adds a new H2 section, [Liquid Presence Verification](http://sandbox.docs.opentrons.com/docs-liquid-presence-detection/v2/pipettes/loading.html#liquid-presence-verification) to the page, Loading Pipettes (`loading.rst`). Includes two H3 subsections:

- [Enable LPV Globally](http://sandbox.docs.opentrons.com/docs-liquid-presence-detection/v2/pipettes/loading.html#enable-lpv-globally)
- [Turning LPV Off and On](http://sandbox.docs.opentrons.com/docs-liquid-presence-detection/v2/pipettes/loading.html#turing-lpv-off-and-on)

**New**
Adds two new H2 sections to the page, Liquid Control (`liquids.rst`):

- [Detect Liquids](http://sandbox.docs.opentrons.com/docs-liquid-presence-detection/v2/basic_commands/liquids.html#detect-liquids)
- [Require Liquids](http://sandbox.docs.opentrons.com/docs-liquid-presence-detection/v2/basic_commands/liquids.html#require-liquids)

**Revisions/Additions**
Small changes to docstrings for:

- `liquid_presence_detection`, an argument of [load_instrument](http://sandbox.docs.opentrons.com/docs-liquid-presence-detection/v2/new_protocol_api.html?highlight=load_instrument#opentrons.protocol_api.ProtocolContext.load_instrument)
- [detect_liquid_presence](http://sandbox.docs.opentrons.com/docs-liquid-presence-detection/v2/new_protocol_api.html?highlight=detect_liquid_presence) in v2 Reference.
- [require_liquid_presence](http://sandbox.docs.opentrons.com/docs-liquid-presence-detection/v2/new_protocol_api.html?highlight=require_liquid_presence)

# Review requests

Take a look and or before adding others.

# Risk assessment

Low. Docs changes only.


[RTC-148]: https://opentrons.atlassian.net/browse/RTC-148?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

jwwojak committed:
Test push

Aa a reminder of how to do this. It's been a while.

jwwojak committed:
Test push

Aa a reminder of how to do this. It's been a while.

jwwojak committed:
Removing test text, not important

jwwojak committed:
Removing test text, not important

jwwojak committed:
Edit LPD argument definition.

Make it clear this is for Flex only and include the version.

jwwojak committed:
Edit LPD argument definition.

Make it clear this is for Flex only and include the version.

jwwojak committed:
H2 LLD section summary

jwwojak committed:
H2 LLD section summary

jwwojak committed:
LLD intro and global method

jwwojak committed:
LLD intro and global method

jwwojak committed:
Adding link to drop_tip

jwwojak committed:
Adding link to drop_tip

jwwojak committed:
Turning global LLD off

jwwojak committed:
Turning global LLD off

jwwojak committed:
End of day save

Deve environment conked out.

jwwojak committed:
End of day save

Deve environment conked out.

jwwojak committed:
Update loading.rst

jwwojak committed:
Update loading.rst

jwwojak committed:
Fix shortened API reference

Forgot ~.

jwwojak committed:
Fix shortened API reference

Forgot ~.

jwwojak committed:
Fixing code formatting

jwwojak committed:
Fixing code formatting

jwwojak committed:
End 'o day commit, will shorten code example

jwwojak committed:
End 'o day commit, will shorten code example

jwwojak committed:
Main part decent draft

This is the bulk or main part of LLD. The smaller methods, detect_liquid_presence and require_liquid_presence, are not here or in this draft. Those will go in the building block commands doc.

jwwojak committed:
Main part decent draft

This is the bulk or main part of LLD. The smaller methods, detect_liquid_presence and require_liquid_presence, are not here or in this draft. Those will go in the building block commands doc.

jwwojak committed:
Small changes

jwwojak committed:
Small changes

jwwojak committed:
Removed text, added version macro

jwwojak committed:
Removed text, added version macro

jwwojak committed:
Saving because I haven't yet

jwwojak committed:
Saving because I haven't yet

jwwojak committed:
Building block commands: detect & require

jwwojak committed:
Building block commands: detect & require

jwwojak committed:
Misspellings in commented text

Fixing because it raises warnings during docs builds.

jwwojak committed:
Misspellings in commented text

Fixing because it raises warnings during docs builds.

jwwojak committed:
More revisions

Name changes and stubs.

jwwojak committed:
More revisions

Name changes and stubs.

jwwojak commented:
Didn't check w/ Black before opening PR.

aaron-kulkarni commented on api/docs/v2/pipettes/loading.rst:
```diff
@@ -215,3 +215,66 @@ Another example is a Flex protocol that uses a waste chute. Say you want to only
+Liquid Presence Verification
+============================
+
+Liquid Presence Verification (LPV) is a pressure-based feature that gives all Opentrons pipettes the ability to detect the presence or absence of a liquid in a well or reservoir. You can use LPV to identify, avoid, and recover from liquid-related protocol errors or just check for the presence or absence of a fluid---all with or without interrupting a protocol run. You can activate LPV globally fo ...
```
Don't know if I'm qualified to review PR's but...
LPV will only work for the Flex. The OT-2 pipettes don't have the necessary pressure sensors.

jwwojak replied:
Oh, drat. That's right. My mistake. That paragraph was a last minute change. Needs to specify Flex-only.

jwwojak committed:
Updating LPV intro - stress Flex only

jwwojak committed:
Updating LPV intro - stress Flex only

jwwojak committed:
detect_ and require_ revisions

Changes to descriptions in Building Block Commands and in the docstring references.

jwwojak committed:
detect_ and require_ revisions

Changes to descriptions in Building Block Commands and in the docstring references.

jwwojak committed:
Reviewer changes and feature name change

Changes from Ed's recommendations (Slack) and change name to "Liquid Presence Detection" from "Liquid Presence Verification."

Some outstanding issues: code sample. Will get help w/ this.

jwwojak committed:
Reviewer changes and feature name change

Changes from Ed's recommendations (Slack) and change name to "Liquid Presence Detection" from "Liquid Presence Verification."

Some outstanding issues: code sample. Will get help w/ this.

jwwojak committed:
Small intro changes and to some how it works text

jwwojak committed:
Small intro changes and to some how it works text

jwwojak committed:
Explain using LLD requires extra time to execute

jwwojak committed:
Explain using LLD requires extra time to execute

aaron-kulkarni commented on api/docs/v2/basic_commands/liquids.rst:
```diff
@@ -258,3 +258,28 @@ This example aspirates enough air to fill the remaining volume in a pipette::
+Detect Liquids
+==============
+
+The :py:meth:`.InstrumentContext.detect_liquid_presence` method tells a Flex pipette to check for the presence of a liquid in a wellplate or reservoir. This method returns ``True`` if the pressure sensors in a pipette detect a liquid and ``False`` if liquid if the sensors do not. Detection takes place during aspiration. As the Flex pipette detects a liquid, it stops, raises itself above the liqui ...
```
_"Detection takes place during aspiration. As the Flex pipette detects a liquid, it stops, raises itself above the liquid's surface, and then resumes aspiration."_

Detections do take place during an aspiration, but the function that does the automatic detection during an aspiration is `require_liquid_presence`, not `detect_liquid_presence`. 

Also, both require/detect can be called entirely separately from aspirations. You can call `detect_liquid_presence` on 100 wells without ever doing an aspiration. Because of this, it might be a good idea to keep explanations of automatic detection during aspirations separate from explanations of these two functions.(I'm not aware of how docs are typically written though so I could be very off here)

jwwojak replied:
@aaron-kulkarni Thanks for the clarifications. These are some important distinctions.

- `detect_liquid_presence` and `require_liquid_presence` can be called separately from aspiration.
- function that does the automatic detection during an aspiration is `require_liquid_presence`, not `detect_liquid_presence`.

y3rsh commented:
Sorry for all the broken comments on here.  You can delete them.  If you pull edge into this branch everything is fixed and you wont be hassled by the comments.

jwwojak committed:
Not ready yet. More TK Monday.

I'm a bit burned out. Saving to save.

jwwojak committed:
Not ready yet. More TK Monday.

I'm a bit burned out. Saving to save.

jwwojak committed:
Shorten code sample, revise explanation

jwwojak committed:
Shorten code sample, revise explanation

jwwojak committed:
Merge branch 'docs-liquid-presence-detection' of https://github.com/Opentrons/opentrons into docs-liquid-presence-detection

ecormany commented on api/docs/v2/pipettes/loading.rst:
```diff
@@ -215,3 +215,74 @@ Another example is a Flex protocol that uses a waste chute. Say you want to only
+Liquid Presence Detection
+=========================
+
+Liquid Presence Detection is a pressure-based feature that allows Opentrons Flex pipettes to detect the presence or absence of liquids in a well, reservoir, or other container. Liquid Presence Detection gives you the ability to identify, avoid, and recover from liquid-related protocol errors. You can enable this feature for an entire protocol run or toggle it on and off as required. Liquid Presen ...
```
This feature name should be lowercase: "liquid presence detection".

ecormany commented on api/docs/v2/pipettes/loading.rst:
```diff
@@ -215,3 +215,74 @@ Another example is a Flex protocol that uses a waste chute. Say you want to only
+Liquid Presence Detection is a pressure-based feature that allows Opentrons Flex pipettes to detect the presence or absence of liquids in a well, reservoir, or other container. Liquid Presence Detection gives you the ability to identify, avoid, and recover from liquid-related protocol errors. You can enable this feature for an entire protocol run or toggle it on and off as required. Liquid Presen ...
+
+.. note::
+    If your protocol uses :ref:`partial tip pickup <partial-tip-pickup>`, the pressure sensors for the Flex 8-Channel pipette are on channels 1 and 8. For the Flex 96-Channel Pipette, the pressure sensors are on channels 1 and 96.
```
```suggestion
    If your protocol uses :ref:`partial tip pickup <partial-tip-pickup>`, the pressure sensors for the Flex 8-Channel pipette are on channels A1 and H1. For the Flex 96-Channel Pipette, the pressure sensors are on channels A1 and H12.
```
Matches how we refer to these channels/nozzles for partial tip pickup.

ecormany commented on api/docs/v2/pipettes/loading.rst:
```diff
@@ -215,3 +215,74 @@ Another example is a Flex protocol that uses a waste chute. Say you want to only
+.. Ed: Enabling instead of Enable?
+
+Enable Liquid Presence Detection
+--------------------------------
```
Yes, I would go with the gerund form to match other headers. It would also be good to shorten these headers. They're at the fourth level of the side nav, so they can rely on the context provided by their parent section.

ecormany commented on api/docs/v2/pipettes/loading.rst:
```diff
@@ -215,3 +215,74 @@ Another example is a Flex protocol that uses a waste chute. Say you want to only
+Enable Liquid Presence Detection
+--------------------------------
+
+The easiest, and recommended, way to use Liquid Presence Detection is by adding the optional Boolean argument, ``liquid_presence_detection=True`` to :py:meth:`.ProtocolContext.load_instrument` in your protocol. When ``True``, the robot will check for liquid on every aspiration. You can also turn this feature off and back on again later in a protocol. This example adds Liquid Presence Detection to ...
```
- Not recommended, because of how time-consuming it is.
- Remove comma after "argument" since the monospace formatting does the work (and this isn't an appositive).
- Last sentence, probably better to say "enables…for" instead of "adds…to"

ecormany commented on api/docs/v2/pipettes/loading.rst:
```diff
@@ -215,3 +215,74 @@ Another example is a Flex protocol that uses a waste chute. Say you want to only
+.. note::
+    If your protocol uses :ref:`partial tip pickup <partial-tip-pickup>`, the pressure sensors for the Flex 8-Channel pipette are on channels 1 and 8. For the Flex 96-Channel Pipette, the pressure sensors are on channels 1 and 96.
+
+.. versionadded:: 2.20
```
Can remove this `versionadded` notice since no code shown yet in this section.

jwwojak replied:
@ecormany I was talking to Anthony, going back to that conversation about about LLD and partial tip pickup. He seems concerned that docs should include text reminding/warning users that the other channels can't sense liquid. Even though it is obvious, I was going to add another sentence along the lines of "these are the only pipette channels that can do LLD" or "When using partial tip pickup, other pipette channels cannot detect the presence/absence of a liquid." 

Need some help w/ guidance that points in 2 different directions–remove or keep and enhance.

ecormany replied:
Let's include a brief description of this behavior in the API Reference entries for `detect_liquid_presence()` and `require_liquid_presence()`

jwwojak replied:
Making a note to self as a reminder:

- Remove the note text
- Put it in the reference
- Add the info about there's no sensors on all the channels, partial tip_won't_ detect on those non-sensor channels.

ecormany commented on api/docs/v2/pipettes/loading.rst:
```diff
@@ -215,3 +215,74 @@ Another example is a Flex protocol that uses a waste chute. Say you want to only
+
+Let's take a look at how all this works. First, tell the robot to pick up a new, clean tip::
+    
+    pipette.pick_up_tip(tiprack2)
+
+Next, tell the robot to aspirate and dispense some liquid from the reservoir::
+
+    pipette.aspirate(100, reservoir["A1"])
+    pipette.dispense(100, plate["A1"])
```
I don't think we gain anything from splitting up these code blocks. It makes it seem like the pickup action does something differently. It would be best to combine them and annotate where LPD happens with a comment.

Also, these don't match the pipette from before, which was loaded as `right`.

jwwojak replied:
Oops. I see it now. Rewriting, fixing.

ecormany commented on api/docs/v2/pipettes/loading.rst:
```diff
@@ -215,3 +215,74 @@ Another example is a Flex protocol that uses a waste chute. Say you want to only
+    
+    pipette.pick_up_tip(tiprack2)
+    pipette.liquid_presence_detection=False  #Liquid Presence Detection off
+    pipette.aspirate(100, reservoir["A2"])
```
```suggestion
    pipette.pick_up_tip()
    pipette.liquid_presence_detection=False  # liquid presence detection off
    pipette.aspirate(100, reservoir["A2"])  # aspirates immediately
```

no need to complicate with pickup location.

ecormany commented on api/docs/v2/pipettes/loading.rst:
```diff
@@ -215,3 +215,74 @@ Another example is a Flex protocol that uses a waste chute. Say you want to only
+
+Going forward, the pipette will not perform a liquid check until you turn this feature back on. 
+
+To reactivate, add ``pipette.liquid_presence_detection=True`` at the point later in a protocol where it needs to be enabled, usually between picking up a new tip and aspirating a liquid.
```
```suggestion
To reactivate, set ``liquid_presence_detection=True`` at the point later in the protocol where it needs to be enabled, usually between picking up a new tip and aspirating a liquid.
```

ecormany commented on api/docs/v2/pipettes/loading.rst:
```diff
@@ -215,3 +215,74 @@ Another example is a Flex protocol that uses a waste chute. Say you want to only
+
+    pipette.pick_up_tip(tiprack2)
+    pipette.liquid_presence_detection=True  #Liquid Presence Detection on again
+    pipette.aspirate(100, reservoir["A3"])
```
```suggestion
    pipette.pick_up_tip()
    pipette.liquid_presence_detection=True  # liquid presence detection on again
    pipette.aspirate(100, reservoir["A3"])  # detects liquid before aspirating
```

ecormany commented on api/docs/v2/pipettes/loading.rst:
```diff
@@ -215,3 +215,74 @@ Another example is a Flex protocol that uses a waste chute. Say you want to only
+
+The robot will continue to check for a liquid until this feature is disabled again, or an empty well is detected (and the robot raises an error), or the protocol completes.
+
+See also :ref:`detect-liquid-presence` and :ref:`require-liquid-presence`.
```
It would be good to add context to these cross-references by moving them up to line 236 and adding an explanation that you should click through to these other sections if you want to do one-off detection.

jwwojak replied:
Line numbers are slightly different now. Moving content up to recommended location and adding more text about `detect_liquid_presence` and `require_liquid_presence`. Will write about them as standalone methods you can use without aspirating and give a summary of what they do. Methods will be linked to reference.

ecormany commented on api/docs/v2/basic_commands/liquids.rst:
```diff
@@ -258,3 +258,68 @@ This example aspirates enough air to fill the remaining volume in a pipette::
+Detect Liquids
+==============
+
+The :py:meth:`.InstrumentContext.detect_liquid_presence` method tells a Flex pipette to check for the presence of a liquid in a wellplate or reservoir. This method returns ``True`` if the pressure sensors in a pipette detect a liquid and ``False`` if the sensors do not. It will not raise an error or stop your protocol when the robot detects and empty well. Detection takes place during aspiration, ...
```
```suggestion
The :py:meth:`.InstrumentContext.detect_liquid_presence` method tells a Flex pipette to check for the presence of a liquid in a well plate or reservoir. This method returns ``True`` if the pressure sensors in the pipette detect a liquid and ``False`` if the sensors do not. It will not raise an error or stop your protocol when the robot detects an empty well. 

You should generally detect liquid immediately after picking up a tip and before aspirating, but you don't need to aspirate to use ``detect_liquid_presence()``. It's a standalone method that can be called when you only want to detect liquid. 

.. note::
    Accurate liquid detection requires fresh, dry pipette tips.

```

Couple typos and more specific guidance on when is ok/good to use this method.

ecormany commented on api/docs/v2/basic_commands/liquids.rst:
```diff
@@ -258,3 +258,68 @@ This example aspirates enough air to fill the remaining volume in a pipette::
+
+In cases where you need records that are more informative than ``true`` or ``false``, you can write your own code to handle the output of this method. For example, by adding some indexing along with ``for`` and ``while`` loops, your protocol could count how many aspirations a robot performed before a source well ran out of liquid. Here's one way to do this in the ``run():`` function of our Flex : ...
+
+.. code-block:: python
```
This code block needs an overhaul. From most to least important:

- The `for` loop is broken because labware don't have an iterable `items()` method, much less one that has names and wells. 
  - It needs to be `for well in reservoir.wells()` or this will not analyze.
  - The reference to `name` in the comment should be `well.name`.
- Variable case and names don't match, which will lead to Python syntax errors. Variable names should be snake_case when assigned and referenced.
- The tip index is superfluous because it's never referenced in the messages at the end and you could use automatic tip tracking instead.
- General `black`ening
  - Two spaces before comment delimiters
  - `+=` is an assignment operator and should have spaces on either side

jwwojak replied:
We're often not following our own rules re: [comments](https://peps.python.org/pep-0008/#comments).

> Comments should be complete sentences. The first word should be capitalized, unless it is an identifier that begins with a lower case letter (never alter the case of identifiers!).

ecormany commented on api/docs/v2/basic_commands/liquids.rst:
```diff
@@ -258,3 +258,68 @@ This example aspirates enough air to fill the remaining volume in a pipette::
+===============
+
+The :py:meth:`.InstrumentContext.require_liquid_presence` method tells a Flex pipette to check for and require a liquid in a well or reservoir. This method returns ``True`` if the pressure sensors in a pipette detect a liquid and ``False`` if liquid if the sensors do not. When the robot does not detect liquid, it raises an error, stops the protocol to let you resolve the problem, and writes a war ...
+Detection takes place during aspiration, but you don't need to aspirate to use ``require_liquid_presence``. It's a standalone method that can be called when you want to detect liquids only. See also :ref:`lpd`.
```
As above, you should call this with a fresh tip and before aspiration.

And the cross-reference can be strengthened to indicate that it links to instructions on how to perform this automatically for all aspirations on a given pipette.

ecormany commented on api/docs/v2/basic_commands/liquids.rst:
```diff
@@ -258,3 +258,68 @@ This example aspirates enough air to fill the remaining volume in a pipette::
+
+.. code-block:: python
+
+    pipette.require_liquid_presence()
```
Needs a well argument.

ecormany commented on api/docs/v2/basic_commands/liquids.rst:
```diff
@@ -258,3 +258,68 @@ This example aspirates enough air to fill the remaining volume in a pipette::
+
+.. code-block:: python
+
+    pipette.detect_liquid_presence()
```
Needs a well argument.

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -2120,7 +2120,7 @@ def configure_nozzle_layout(  # noqa: C901
     @requires_version(2, 20)
     def detect_liquid_presence(self, well: labware.Well) -> bool:
-        """Check if there is liquid in a well.
+        """Checks if there is liquid in a well. Will not raise an error if it does not detect liquid. Flex pipettes only. See :ref:`lpd` and :ref:`detect-liquid-presence`.
```
```suggestion
        """Check if there is liquid in a well. 
        
        Returns ``True`` if liquid is present and ``False`` if liquid is not present. Will not raise an error if it does not detect liquid. Flex pipettes only. See :ref:`detect-liquid-presence`.
```
First sentence of a docstring should stand alone.

Removed xref to global detection, which is more like `require_liquid_presence()`.

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -2130,7 +2130,7 @@ def detect_liquid_presence(self, well: labware.Well) -> bool:
     @requires_version(2, 20)
     def require_liquid_presence(self, well: labware.Well) -> None:
-        """If there is no liquid in a well, raise an error.
+        """Checks for liquid in a well. Raises an error if no liquid is detected. Flex pipettes only. See :ref:`lpd` and :ref:`require-liquid-presence`.
```
```suggestion
        """Check for liquid in a well and raise an error if none is detected. 
        
        When this method raises an error, Flex will offer the opportunity to enter recovery mode. In recovery mode, you can manually add liquid to resolve the error. 
        
        Works with Flex pipettes only. See :ref:`lpd` and :ref:`require-liquid-presence`.
```

Suggesting to differentiate the first lines and explain a bit more about what you can do when an error happens. (For power users, this also implies that you don't need to catch this error in a `try...except` block.)

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -2120,7 +2120,7 @@ def configure_nozzle_layout(  # noqa: C901
-        """Check if there is liquid in a well.
+        """Checks if there is liquid in a well. Will not raise an error if it does not detect liquid. Flex pipettes only. See :ref:`lpd` and :ref:`detect-liquid-presence`.
 
         :returns: A boolean.
```
We could remove line 2125, or move the prose description of the values of true and false here.

jwwojak replied:
Will move the longer, prose description to `:returns:` (line 2125).

ecormany commented on api/src/opentrons/protocol_api/instrument_context.py:
```diff
@@ -2130,7 +2130,7 @@ def detect_liquid_presence(self, well: labware.Well) -> bool:
-        """If there is no liquid in a well, raise an error.
+        """Checks for liquid in a well. Raises an error if no liquid is detected. Flex pipettes only. See :ref:`lpd` and :ref:`require-liquid-presence`.
 
         :returns: None.
```
Remove line 2135 or style as `None`

jwwojak replied:
Not following here. I think you mean do the same thing as in `detect_liquid_presence` above.

ecormany commented on api/src/opentrons/protocol_api/protocol_context.py:
```diff
@@ -906,7 +906,9 @@ def load_instrument(
                              replace an instrument in the middle of a protocol being run
                              from the Opentrons App or touchscreen.
-        :param bool liquid_presence_detection: If ``True``, enable liquid presence detection for instrument. Only available on Flex robots in API Version 2.20 and above.
+        :param bool liquid_presence_detection: If ``True``, enables liquid presence detection for Flex 1-, 8-, or 96-channel pipettes. See :ref:`lpd`.
```
```suggestion
        :param bool liquid_presence_detection: If ``True``, enable liquid presence detection for Flex 1-, 8-, or 96-channel pipettes. See :ref:`lpd`.
```
Should be imperative like other param descriptions.

ecormany reviewed (changes requested):
Some changes to make. Semantic changes are around being clear about exactly when LLD happens, the difference between detecting and requiring, and the strong link between the require standalone method and global detection (naming things is hard, so it goes). Syntactic changes are to get the big code block working — get engineer support as needed. And stylistic changes are all to be blamed on the Pythonistas 🐍 🧐

ecormany committed:
add behavior in simulation to API ref entries

jwwojak committed:
Apply suggestions from code review

Committing recommended changes. Will continue to work on others.

Co-authored-by: Ed Cormany <edward.cormany@opentrons.com>

jwwojak committed:
Revisions from reviewers

Made changes based on Ed's comments and our meeting today. Got many, but not all. More to come.

jwwojak committed:
Another round of reviewer comments and changes

jwwojak committed:
Removed code sample, revise method text

jwwojak committed:
More reviewer changes

-- removed code sample
-- added text about pressure sensor locations
-- ready for review again

ecormany committed:
review and edits

ecormany reviewed (approved):
Went through final changes with Joe. :shipit:
